src/rgw/rgw_auth_s3.cc

```diff
@@ -700,13 +703,18 @@ std::string get_v4_string_to_sign(CephContext* const cct,
                                   const sha256_digest_t& canonreq_hash)
 {
   const auto hexed_cr_hash = buf_to_hex(canonreq_hash);
-  const auto string_to_sign = std::string()
+
+  /* We aren't adding the extra byte for the terminating 0 as it would be
+   * annihilated by the "-1" coming from hexed_cr_hash. */
```

Comment:
@rzarzynski the new code looks good to me (great work!) Bear in mind I am also a bit concerned about the new code revamp, it adds lots of c++ stuff and while it adds great structure, cleanup, etc. I have the feeling we are losing control over simple things as the variable life scope, string lengths, etc.

Maybe we could stick to c++ advantages on the general structure and keeping the string handling and low level stuff under control with simple and explicit allocations in stack/heap where we can count bytes, knowing the pointer/ref is valid, etc. Not sure though. Just a suggestion to avoid possible off-by-one and friends.